SwiftCompilerSources/Sources/SIL/Value.swift

```diff
@@ -26,9 +26,14 @@ public protocol Value : AnyObject, CustomStringConvertible {
   
   /// The block where the value is defined.
   ///
-  /// It's not legal to get the definingBlock of an `Undef` value.
+  /// It's not legal to get the parentBlock of an `Undef` value or an instruction in a global initializer.
```

Comment:
> parentBlock of an `Undef` value
Is this consistent with 
```
  if (auto *undef = dyn_cast<SILUndef>(this)) {
    // By convention, undefs are considered to be defined at the entry of the function.
    return undef->getParent()->getEntryBlock();
  }
```
?

Reply:
looks like an oversight to me too

Reply:
Question. @eeckstein I thought the whole point of the global variable initializer design is that it reuses a SILBasicBlock... so an instruction in one should be able to return its parent block. Are we just not allowing that due to the way we want the API to be used? Just would like to understand.

Reply:
yes, I removed this comment

Reply:
> I thought the whole point of the global variable initializer design is that it reuses a SILBasicBlock

Not exactly. Using a SILBasicBlock to store initializer instructions is just an implementation detail. Conceptually global init instructions don't have a parent block. Unfortunately this "internal" block gets exposed by `getParent`.

Reply:
> Conceptually global init instructions don't have a parent block.

That is not a reasonable or viable concept. If that were true, then the SILInstruction interface is itself incorrect, since parentBlock is non-optional.  And all utilities that operate on a SIL Instruction are buggy. And every other engineer who works on the SIL codebase will have a fundamental misconception that instructions are in blocks, because that's how every other IR works and that's how SIL happens to work > 99% of the time.

It is viable, and reasonable to say that global initializers are single-block "SIL function bodies". Where a SIL function body is simply a single-entry list of blocks. If you want to do anything with a SIL function body that is not supported by globals initializers, then you should need to explicitly downcast to from `SILFunctionBody` to `SILFunction`, which is really an AST concept that mostly has nothing to do with SIL.

Reply:
A statically initialized global is not something where "instructions" are "executed" in some sequential order. It is a piece of memory in the data section containing some values. Therefore viewing the static initializer as a function is simply wrong.
Rather than a _list_ of instructions, a static initializer is represented as a tree-like data structure, similar to llvm's Constant. The root of the tree is the global's `staticInitValue`. The nodes are SingleValueInstructions, where the operands lead to the next nodes in the tree (actually: DAG). Although the GlobalVariable has APIs to iterate over all instructions, this is more like a "visit all nodes in the tree".

Would it be cleaner to define separate node data structures, which would be named `*Const` instead of `*Inst`? Yes
Would it be worth the effort? No
Would it make things simpler? No (not only that we need to replicate all the instruction classes, we also wouldn't be able to use the cloner to translate between instructions <-> constants)
Is it a problem that the static initializer contains instructions with no parent block/function? No, because all the passes and utilities will never come in touch with static initializer instructions.

Reply:
There is a good argument for not implementing separate ConstExpr types. But nothing about constant evaluation forces you to violate fundamental SIL types. All the rest of SIL's ConstExpr evaluation is happy with the standard SIL types. If they had a return instruction, then static initializers could have a SIL function body without changing any semantics. It would be very interesting for me to see which SILFunction APIs are categorically invalid for initializers. The hardest part about changing this would be either duplicating many SILFunction APIs or downcasting to SILFunction for the ones that aren't duplicated.

Reply:
@eeckstein @atrick suggestion. I wonder if we could create two separate types. Maybe do something with templates? Then we can type alias the two. Just trying to explore the state space.

Reply:
There's an important principle in software design that an apparent invariant is a problem when it is not actually invariant. The closer that apparent invariant gets to being invariant, without being actually invariant, the bigger the problem. If the apparent invariant proves to be false quickly as soon as you test it, the problem is minor. But each time people can act as if it is invariant without being corrected, the incorrect belief is reinforced, and the bugs harder to find.

So this case, in principle, seems as bad as you can get. But, as far as I know, the only problem it causes in practice  is that people working on SIL simply have no way of knowing where it is valid to use `parentFunction`. We certainly have much bigger problems to worry about.